src/NuGet.Core/NuGet.Repositories/NuGetv3LocalRepository.cs

```diff
@@ -2,22 +2,37 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using NuGet.Packaging;
 using NuGet.Versioning;
 
 namespace NuGet.Repositories
 {
+    /// <summary>
+    /// Caches package info from the global packages folder in memory.
+    /// Packages not yet in the cache will be retrieved from disk.
+    /// </summary>
     public class NuGetv3LocalRepository
     {
-        private readonly Dictionary<string, IEnumerable<LocalPackageInfo>> _cache = new Dictionary<string, IEnumerable<LocalPackageInfo>>(StringComparer.OrdinalIgnoreCase);
+        private readonly ConcurrentDictionary<string, IEnumerable<LocalPackageInfo>> _cache
+            = new ConcurrentDictionary<string, IEnumerable<LocalPackageInfo>>(StringComparer.OrdinalIgnoreCase);
```

Comment:
Sort of, in a lot of cases here it just tracks that a package doesn't exist on disk and that it needs to be downloaded. Storing the nuspecs could start to build up, but I think that needs to be solved along with project.json and project.lock.json caching, the lock file in particular is massively larger than the nuspecs here.